src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs

```diff
@@ -36,6 +42,208 @@ internal IntPtr[] GetWaitedHandleArray(int requiredCapacity)
+        }
+
+        private static SafeWaitHandle GetOSHandleForCurrentThread()
+        {
+            SafeWaitHandle threadHandle;
+            if (Interop.mincore.GetCurrentThreadHandle(out threadHandle))
+            {
+                return threadHandle;
+            }
+
+            // Throw an ApplicationException for compatibility with CoreCLR
+            var ex = new ApplicationException();
+            ex.SetErrorCode((int)Interop.mincore.GetLastError());
+            throw ex;
+        }
+
+        private static ThreadPriority MapFromOSPriority(OSThreadPriority priority)
+        {
+            switch (priority)
+            {
+                case OSThreadPriority.Idle:
+                case OSThreadPriority.Lowest:
+                    return ThreadPriority.Lowest;
+
+                case OSThreadPriority.BelowNormal:
+                    return ThreadPriority.BelowNormal;
+
+                case OSThreadPriority.Normal:
+                    return ThreadPriority.Normal;
+
+                case OSThreadPriority.AboveNormal:
+                    return ThreadPriority.AboveNormal;
+
+                case OSThreadPriority.Highest:
+                case OSThreadPriority.TimeCritical:
+                    return ThreadPriority.Highest;
+
+                case OSThreadPriority.ErrorReturn:
+                    Debug.Fail("GetThreadPriority failed");
+                    return ThreadPriority.Normal;
+
+                default:
+                    return ThreadPriority.Normal;
+            }
+        }
+
+        private static OSThreadPriority MapToOSPriority(ThreadPriority priority)
+        {
+            switch (priority)
+            {
+                case ThreadPriority.Lowest:
+                    return OSThreadPriority.Lowest;
+
+                case ThreadPriority.BelowNormal:
+                    return OSThreadPriority.BelowNormal;
+
+                case ThreadPriority.Normal:
+                    return OSThreadPriority.Normal;
+
+                case ThreadPriority.AboveNormal:
+                    return OSThreadPriority.AboveNormal;
+
+                case ThreadPriority.Highest:
+                    return OSThreadPriority.Highest;
+
+                default:
+                    Environment.FailFast("Unreached");
+                    return OSThreadPriority.Normal;
+            }
+        }
+
+        private bool SetPriority(ThreadPriority priority)
+        {
+            if (_osHandle.IsInvalid)
+            {
+                return true;
+            }
+            return Interop.mincore.SetThreadPriority(_osHandle, (int)MapToOSPriority(priority));
+        }
+
+        /// <summary>
+        /// Checks if the underlying OS thread has finished execution.
+        /// </summary>
+        private bool HasFinishedExecution()
+        {
+            SafeWaitHandle waitHandle = _osHandle;
+            uint result;
+
+            waitHandle.DangerousAddRef();
```

Comment:
Is there a good reason for the manual marshalling here and not just make the argument of WaitForSingleObject to be SafeHandle?